### PR TITLE
Coh 155 draft relisting

### PR DESCRIPTION
**Pull request by johnhosolirius**

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/coh/controller/RelistingController.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    @GetMapping
+    public ResponseEntity retrieveRelisting(@PathVariable UUID onlineHearingId) {
+        Optional<OnlineHearing> optionalOnlineHearing = onlineHearingService.retrieveOnlineHearing(onlineHearingId);
+        if (!optionalOnlineHearing.isPresent()) {
+            return ResponseEntity.status(HttpStatus.NOT_FOUND).body("Not found");
+        }
+        OnlineHearing onlineHearing = optionalOnlineHearing.get();
+        String reason = onlineHearing.getRelistReason();
+        RelistingState state = onlineHearing.getRelistState();
+        Relisting relisting = new Relisting(reason, state);
```
Comment (johnhosolirius) [inline src/main/java/uk/gov/hmcts/reform/coh/controller/RelistingController.java]: This return 200 even if there's no relist reason

File: src/main/java/uk/gov/hmcts/reform/coh/controller/RelistingController.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        if (!optionalOnlineHearing.isPresent()) {
+            return ResponseEntity.status(HttpStatus.NOT_FOUND).body("Not found");
+        }
+
+        OnlineHearing onlineHearing = optionalOnlineHearing.get();
+
+        if (onlineHearing.getRelistState() == RelistingState.ISSUED) {
+            return ResponseEntity.status(HttpStatus.CONFLICT).body("Already issued");
+        }
+
```
Comment (johnhosolirius) [inline src/main/java/uk/gov/hmcts/reform/coh/controller/RelistingController.java]: 1. If the relist is issued to the appellant, then the online hearing also has to be updated
2. I thought we agree to queue the request to follow the the same pattern as issuing of questions/decisions etc

File: src/test/java/uk/gov/hmcts/reform/coh/controller/RelistingControllerTest.java [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+            .andExpect(jsonPath("$.state", is("DRAFTED")));
+    }
+
+    @Test
+    public void creatingRelistingRequiresStateField() throws Exception {
+        String request = "{}";
+        mockMvc.perform(post(pathToExistingOnlineHearing).content(request).contentType(APPLICATION_JSON))
+            .andExpect(status().isUnprocessableEntity());
+    }
+}
```
Comment (johnhosolirius) [inline src/test/java/uk/gov/hmcts/reform/coh/controller/RelistingControllerTest.java]: Write Cucumber tests

File: src/main/java/uk/gov/hmcts/reform/coh/controller/RelistingController.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+import java.util.Optional;
+import java.util.UUID;
+import javax.validation.Valid;
+
+@RestController
+@RequestMapping(
+    value = "/continuous-online-hearings/{onlineHearingId}/relist",
+    produces = MediaType.APPLICATION_JSON_VALUE
+)
```
Comment (johnhosolirius) [inline src/main/java/uk/gov/hmcts/reform/coh/controller/RelistingController.java]: Missing Swagger docs

File: src/main/java/uk/gov/hmcts/reform/coh/controller/onlinehearing/Relisting.java [hunk 5]
```diff
@@ -0,0 +1,21 @@
+package uk.gov.hmcts.reform.coh.controller.onlinehearing;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import uk.gov.hmcts.reform.coh.domain.RelistingState;
+
+import javax.validation.constraints.NotNull;
+
+public class Relisting {
```
Comment (johnhosolirius) [inline src/main/java/uk/gov/hmcts/reform/coh/controller/onlinehearing/Relisting.java]: Don't think the Relisting members should be public

File: src/main/resources/db/changelog/db.changelog-diff-20181001-1303.yaml [hunk 6]
```diff
@@ -0,0 +1,12 @@
+databaseChangeLog:
```
Comment (johnhosolirius) [inline src/main/resources/db/changelog/db.changelog-diff-20181001-1303.yaml]: You need to add this changelog to the master change log file

File: src/main/java/uk/gov/hmcts/reform/coh/controller/RelistingController.java (hunk 1 above)
Comment (KarolJastrzebski) [reply src/main/java/uk/gov/hmcts/reform/coh/controller/RelistingController.java]: Yes, reason is optional.

File: src/test/java/uk/gov/hmcts/reform/coh/controller/RelistingControllerTest.java (hunk 3 above)
Comment (KarolJastrzebski) [reply src/test/java/uk/gov/hmcts/reform/coh/controller/RelistingControllerTest.java]: Thanks, I'll add some in the next commits.

File: src/main/java/uk/gov/hmcts/reform/coh/controller/RelistingController.java (hunk 4 above)
Comment (KarolJastrzebski) [reply src/main/java/uk/gov/hmcts/reform/coh/controller/RelistingController.java]: Thanks, I'll add these in the upcoming commits.

File: src/main/java/uk/gov/hmcts/reform/coh/controller/onlinehearing/Relisting.java (hunk 5 above)
Comment (KarolJastrzebski) [reply src/main/java/uk/gov/hmcts/reform/coh/controller/onlinehearing/Relisting.java]: It is still a data structure, whether it uses setters/getters or public fields.

The `public final` approach doesn't require so much boilerplate code. I am happy to change if missed anything valuable.

File: src/main/resources/db/changelog/db.changelog-diff-20181001-1303.yaml (hunk 6 above)
Comment (KarolJastrzebski) [reply src/main/resources/db/changelog/db.changelog-diff-20181001-1303.yaml]: Thanks, added.

File: src/test/java/uk/gov/hmcts/reform/coh/controller/RelistingControllerTest.java [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+
+    @MockBean
+    private OnlineHearingStateService onlineHearingStateService;
+
+    @Autowired
+    private MockMvc mockMvc;
+    private OnlineHearing onlineHearing;
+    private String pathToExistingOnlineHearing;
+    private String pathToNonExistingOnlineHearing;
+
```
Comment (johnhosolirius) [inline src/test/java/uk/gov/hmcts/reform/coh/controller/RelistingControllerTest.java]: Any reason why you can't use JsonUtils?

File: src/test/java/uk/gov/hmcts/reform/coh/controller/RelistingControllerTest.java [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+    @Autowired
+    private ObjectMapper objectMapper;
+
+    @Before
+    public void setup() {
+        onlineHearing = OnlineHearingEntityUtils.createTestOnlineHearingEntity();
+        when(onlineHearingService.retrieveOnlineHearing(onlineHearing.getOnlineHearingId()))
+            .thenReturn(Optional.of(onlineHearing));
+
```
Comment (johnhosolirius) [inline src/test/java/uk/gov/hmcts/reform/coh/controller/RelistingControllerTest.java]: Use CohUriBuilder

Comment (KarolJastrzebski) [reply src/test/java/uk/gov/hmcts/reform/coh/controller/RelistingControllerTest.java]: Thanks, done.

File: src/main/java/uk/gov/hmcts/reform/coh/controller/RelistingController.java (hunk 2 above)
Comment (KarolJastrzebski) [reply src/main/java/uk/gov/hmcts/reform/coh/controller/RelistingController.java]: 1. Online hearing's state is updated right after Relisting's state is changed to `ISSUED`.
2. This feature is coming in the next commits.
3. Will do that after getting feature from 2. point up and running.
4. Reason was supposed to be optional for re-listing or have I missed the spot?
5. Validation is provided by Jackson with `@Valid` annotation.
6. Thanks, I'll add this.
7. Thanks, I'll add this.
8. That reminds me about adding created and updated fields to the response.

File: src/test/java/uk/gov/hmcts/reform/coh/controller/RelistingControllerTest.java (hunk 7 above)
Comment (KarolJastrzebski) [reply src/test/java/uk/gov/hmcts/reform/coh/controller/RelistingControllerTest.java]: Done.

Comment (johnhosolirius) [reply src/test/java/uk/gov/hmcts/reform/coh/controller/RelistingControllerTest.java]: I meant JsonUtils.toObjectFromJson

File: src/main/java/uk/gov/hmcts/reform/coh/controller/onlinehearing/Relisting.java (hunk 5 above)
Comment (johnhosolirius) [reply src/main/java/uk/gov/hmcts/reform/coh/controller/onlinehearing/Relisting.java]: Exposing the inner members of a class doesn't sit right with me. You can always to auto code generator or Lombok?

File: src/main/java/uk/gov/hmcts/reform/coh/controller/RelistingController.java (hunk 2 above)
Comment (KarolJastrzebski) [reply src/main/java/uk/gov/hmcts/reform/coh/controller/RelistingController.java]: - `ISSUE_PENDING` state has been implemented so that re-listing will be issued only after successful delivery of the notification.
- Setting directly online hearing as re-listed when updating is now removed.
- Both fields `created` and `updated` are added.

File: src/aat/java/uk/gov/hmcts/reform/coh/functional/bdd/steps/OnlineHearingSteps.java [hunk 9]
```diff
@@ -310,6 +313,63 @@ public void theOnlineHearingReasonIsReason(String reason) throws Throwable {
         assertEquals(reason, getOnlineHearingResponse().getRelistReason());
     }
 
+    @When("^(drafting|issuing) the relist$")
```
Comment (johnhosolirius) [inline src/aat/java/uk/gov/hmcts/reform/coh/functional/bdd/steps/OnlineHearingSteps.java]: Wouldn't it be easier if this simply said
the relist state is '<ACTUAL STATE THAT WOULD BE SET>'?

File: src/aat/resources/cucumber/answers.feature [hunk 10]
```diff
@@ -137,9 +137,9 @@ Feature: Answers
     And the put request is sent to issue the question round ' "1" '
     And the notification scheduler runs
     Given a standard update online hearing request
-    And the update online hearing state is continuous_online_hearing_relisted
-    And the relist reason is 'reason'
-    And a PUT request is sent for online hearing
+    And the relist reason is set to 'reason'
```
Comment (johnhosolirius) [inline src/aat/resources/cucumber/answers.feature]: Isn't relist reason optional anyway?

File: src/aat/resources/cucumber/relisting.feature [hunk 11]
```diff
@@ -0,0 +1,18 @@
+Feature: Re-listing online hearing
+
+  Scenario: Drafting a reason for re-listing
+    Given a standard online hearing is created
+    And the relist reason is set to 'Not ready yet'
+    When drafting the relist
+    Then the online hearing state is 'continuous_online_hearing_started'
+    And the relist state should be 'drafted'
+
+  @wip
```
Comment (johnhosolirius) [inline src/aat/resources/cucumber/relisting.feature]: Remove this tag

File: src/aat/resources/cucumber/relisting.feature [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+  Scenario: Submitting re-listing reason
+    Given a standard online hearing is created
+    And the relist reason is set to 'Has to be re-listed.'
+    When issuing the relist
+    And the notification scheduler runs
+    And the online hearing state is refreshed
+    Then the online hearing state is 'continuous_online_hearing_relisted'
+    And the relist state should be 'issued'
```
Comment (johnhosolirius) [inline src/aat/resources/cucumber/relisting.feature]: 1. Also need to test empty reason is acceptable
2. Need to test updating

File: src/main/java/uk/gov/hmcts/reform/coh/Application.java [hunk 13]
```diff
@@ -3,14 +3,23 @@
...
 import springfox.documentation.swagger2.annotations.EnableSwagger2;
 
+import java.time.Clock;
+
 @SpringBootApplication
 @EnableSwagger2
 @EnableAutoConfiguration
 @EnableScheduling
 public class Application {
+
```
Comment (johnhosolirius) [inline src/main/java/uk/gov/hmcts/reform/coh/Application.java]: What's the use of Clock?

File: src/main/java/uk/gov/hmcts/reform/coh/controller/RelistingController.java [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+            return ResponseEntity.status(HttpStatus.NOT_FOUND).body("Not found");
+        }
+
+        if (body.state == RelistingState.ISSUE_PENDING) {
+            return ResponseEntity.badRequest().body("Invalid state");
+        }
+
+        OnlineHearing onlineHearing = optionalOnlineHearing.get();
+
+        Date now = Date.from(clock.instant());
```
Comment (johnhosolirius) [inline src/main/java/uk/gov/hmcts/reform/coh/controller/RelistingController.java]: Again. why clock? Is this to make testing easier?

File: src/main/java/uk/gov/hmcts/reform/coh/controller/RelistingController.java [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+
```
Comment (johnhosolirius) [inline src/main/java/uk/gov/hmcts/reform/coh/controller/RelistingController.java]: If you had separate POST and PUT, the logic below would be a lot less convoluted .

File: src/aat/resources/cucumber/answers.feature (hunk 10 above)
Comment (KarolJastrzebski) [reply src/aat/resources/cucumber/answers.feature]: Yes, it is.

File: src/aat/resources/cucumber/relisting.feature (hunk 12 above)
Comment (KarolJastrzebski) [reply src/aat/resources/cucumber/relisting.feature]: Most of these cases are covered in unit tests.

File: src/main/java/uk/gov/hmcts/reform/coh/Application.java (hunk 13 above)
Comment (KarolJastrzebski) [reply src/main/java/uk/gov/hmcts/reform/coh/Application.java]: It allows for injecting specified time from test-level.

File: src/aat/java/uk/gov/hmcts/reform/coh/functional/bdd/steps/OnlineHearingSteps.java (hunk 9 above)
Comment (KarolJastrzebski) [reply src/aat/java/uk/gov/hmcts/reform/coh/functional/bdd/steps/OnlineHearingSteps.java]: There are different styles of writing BDD scenarios:

A:
```gherkin
Given the relist state is 'drafted'
And the relist reason is 'Whatever'
When a PUT request is sent to online hearing relist
```

B:
```gherkin
When drafting a relist with reason "Whatever"
```

They would probably do the same thing underneath, however, the second one communicates intent without overloading the reader with too much technical details. It is easier to comprehend.
